### ISSUE TITLE
How to freeze parameters with `nnx` and `optax`?

### ISSUE THREAD

**Issue by maxencefaldor** [google/flax#4167]
I would like to know what is the best way to freeze parameters in a model using `nnx` and `optax` (https://flax.readthedocs.io/en/latest/guides/training_techniques/transfer_learning.html#optax-multi-transform).

I think it would be useful to add an example to https://flax.readthedocs.io/en/latest/nnx/index.html.

**Comment by cgarciae** [comment]
Check out [08_save_load_checkpoints.py](https://github.com/google/flax/blob/main/flax/nnx/examples/toy_examples/08_save_load_checkpoints.py) which contains a very simple example of orbax checkpointing. We will have a proper checkpointing guide in the future as we migrate the Linen docs.

**Comment by maxencefaldor** [comment]
Thank you for your response, but I believe there's been a misunderstanding. My question was about freezing parameters for transfer learning, not about checkpointing or saving/loading model states.

**Comment by cgarciae** [comment]
Oh god I'm sorry! I read orbax.

Transfer Learning and Surgery in general is a lot simpler in the new `nnx` API. Here's a small working example:


```python
class Classifier(nnx.Module):
  def __init__(self, embed_dim, num_classes, backbone, rngs):
    self.backbone = backbone
    self.head = nnx.Linear(embed_dim, num_classes, rngs=rngs)

  def __call__(self, x):
    x = self.backbone(x)
    x = self.head(x)
    return x

def load_model():
  return nnx.Linear(784, 1024, rngs=nnx.Rngs(0))

backbone = load_model()
classifier = Classifier(1024, 10, backbone, rngs=nnx.Rngs(1))

# filter to select only Params on head path
head_params = nnx.All(nnx.Param, nnx.PathContains('head'))

optimizer = nnx.Optimizer(
  classifier,
  tx=optax.adamw(3e-4),
  wrt=head_params,  # filter head params
)

# sample train step
@nnx.jit
def train_step(model, optimizer, x, y):
  def loss_fn(model):
    logits = model(x)
    return optax.softmax_cross_entropy_with_integer_labels(logits, y).mean()

  diff_state = nnx.DiffState(0, head_params) # filter head params of the first argument
  grads = nnx.grad(loss_fn, argnums=diff_state)(model)
  optimizer.update(grads)

x = jnp.ones((1, 784))
y = jnp.ones((1,), jnp.int32)
train_step(classifier, optimizer, x, y)
```

**Comment by maxencefaldor** [comment]
Excellent, that's exactly what I was looking for!

**Comment by mmorinag127** [comment]
I also have a similar question about `Adamw`'s mask for weight decay.
How can I specify which parameters are to be applied to the weight decay?
Is there any excellent way like the above?

**Comment by cgarciae** [comment]
@mmorinag127 because the `nnx.Optimizer` wrapper is very generic there is no support for `mask` using filters, you can probably generate a compatible mask from the state e.g.

```python
state = nnx.state(classifier)
mask = create_mask(state) # TODO

optimizer = nnx.Optimizer(
  classifier,
  tx=optax.adamw(3e-4, mask=mask),
  wrt=head_params,  # filter head params
)
```

**Comment by mmorinag127** [comment]
Thanks a lot @cgarciae !!

**Comment by rdarder** [comment]
Can you elaborate on how you'd implement create_mask() in your example @cgarciae ?
Thank you!

**Comment by cgarciae** [comment]
@rdarder most likely just a `nnx.map_state` to booleans:

```python
def create_mask(state):
  def mask_fn(path, leaf):
    masked = ... # TODO
    return masked

  return nnx.map_state(mask_fn, state)
```

**Comment by rdarder** [comment]
That unblocked me, thank you!!
If useful to anyone, below is an example for building a mask. 
What got me stuck in the first place is that the state I got form `nnx.split()` was including `BatchStat` nodes (from the batch norm state) that then `optax.masked()` would not recognize). Filtering to only get `nnx.Params` solved it. 


```python
import optax
import flax.nnx as nnx


class SampleModel(nnx.Module):
    def __init__(self, *, rngs: nnx.Rngs):
        self.target_conv = nnx.Conv(in_features=5, out_features=1, kernel_size=(1, 1), rngs=rngs)
        self.batch_norm = nnx.BatchNorm(num_features=29, use_running_average=False, rngs=rngs)


rngs = nnx.Rngs(0)
model = SampleModel(rngs=rngs)
_, state, _ = nnx.split(model, nnx.Param, ...)

def mask_fn(path, s):
    return path == ('target_conv', 'kernel')


mask = nnx.map_state(mask_fn, state)
assert mask['target_conv']['kernel'] == True
assert mask['target_conv']['bias'] == False
assert mask['batch_norm']['scale'] == False

transformations = optax.chain(
    optax.adam(learning_rate=1e-6),
    optax.masked(optax.keep_params_nonnegative(), mask),
)
optimizer = nnx.Optimizer(model, transformations)
```
